File: swift-sdk-objc-tests/IterableActionRunnerTests.m
//
//  IterableActionRunnerTests.m
//  new_ios_sdk_objcTests
//
//  Created by Victor Babenko on 5/14/18.
//  Copyright © 2018 Iterable. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <OCMock/OCMock.h>

@import IterableSDK;

@interface IterableActionRunnerTests : XCTestCase

@end

@implementation IterableActionRunnerTests

- (void)setUp {
    [super setUp];
    [IterableAPI createSharedInstanceWithApiKey:@"" email:@"" userId:nil launchOptions:nil useCustomLaunchOptions:nil];
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

/* !!!TQM: TODO: fix
- (void)testUrlOpenAction {
    id urlDelegateMock = OCMProtocolMock(@protocol(IterableURLDelegate));
    id applicationMock = OCMPartialMock([UIApplication sharedApplication]);
    IterableAPI.instance.urlDelegate = urlDelegateMock;
    IterableAction *action = [IterableAction actionFromDictionary:@{ @"type": @"openUrl", @"data": @"https://example.com" }];
    [IterableActionRunner executeAction:action];
    
    OCMVerify([urlDelegateMock handleIterableURL:[OCMArg isEqual:[NSURL URLWithString:@"https://example.com"]] fromAction:[OCMArg isEqual:action]]);
    if (@available(iOS 10.0, *)) {
        OCMVerify([applicationMock openURL:[OCMArg any] options:[OCMArg any] completionHandler:[OCMArg any]]);
    } else {
        OCMVerify([applicationMock openURL:[OCMArg any]]);
    }
    [applicationMock stopMocking];
}
*/

/* !!!TQM: TODO: fix
- (void)testUrlHandlingOverride {
    id urlDelegateMock = OCMProtocolMock(@protocol(IterableURLDelegate));
    id applicationMock = OCMPartialMock([UIApplication sharedApplication]);
    if (@available(iOS 10.0, *)) {
        OCMReject([applicationMock openURL:[OCMArg any] options:[OCMArg any] completionHandler:[OCMArg any]]);
    } else {
        OCMReject([applicationMock openURL:[OCMArg any]]);
    }
    OCMStub([urlDelegateMock handleIterableURL:[OCMArg any] fromAction:[OCMArg any]]).andReturn(YES);
    IterableAPI.instance.urlDelegate = urlDelegateMock;
    IterableAction *action = [IterableAction actionFromDictionary:@{ @"type": @"openUrl", @"data": @"https://example.com" }];
    [IterableActionRunner executeAction:action];
    
    [applicationMock stopMocking];
}
*/

- (void)testCustomAction {
    id customActionDelegateMock = OCMProtocolMock(@protocol(IterableCustomActionDelegate));
    IterableAPI.instance.customActionDelegate = customActionDelegateMock;
    IterableAction *action = [IterableAction actionFromDictionary:@{ @"type": @"customActionName" }];
    [IterableActionRunner executeAction:action];
    
    OCMVerify([customActionDelegateMock handleIterableCustomAction:[OCMArg checkWithBlock:^BOOL(IterableAction *action) {
        XCTAssertEqualObjects(action.type, @"customActionName");
        return YES;
    }]]);
}

@end

File: swift-sdk-swift-tests/IterableActionRunnerTests.swift
//
//  IterableActionRunnerTests.swift
//  swift-sdk-swift-tests
//
//  Created by Tapash Majumder on 6/14/18.
//  Copyright © 2018 Iterable. All rights reserved.
//

import XCTest

import OHHTTPStubs

@testable import IterableSDK

class IterableActionRunnerTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
        IterableAPI.initialize(apiKey:"", config: IterableConfig())
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testUrlOpenAction() {
        let urlString = "https://example.com"
        let urlDelegate = MockUrlDelegate(returnValue: false)
        let urlOpener = MockUrlOpener()
        let actionRunner = IterableActionRunner(urlDelegate: urlDelegate, customActionDelegate: nil, urlOpener: urlOpener)
        
        let action = IterableAction.action(fromDictionary: ["type" : "openUrl", "data" : urlString])!
        _ = actionRunner.execute(action: action, from: .push)
        
        if #available(iOS 10.0, *) {
            XCTAssertEqual(urlOpener.ios10OpenedUrl?.absoluteString, urlString)
            XCTAssertNil(urlOpener.preIos10openedUrl)
        } else {
            XCTAssertEqual(urlOpener.preIos10openedUrl?.absoluteString, urlString)
            XCTAssertNil(urlOpener.ios10OpenedUrl)
        }
    }
    
    func testUrlHandlingOverride() {
        let urlString = "https://example.com"
        let urlDelegate = MockUrlDelegate(returnValue: true)
        let urlOpener = MockUrlOpener()
        let actionRunner = IterableActionRunner(urlDelegate: urlDelegate, customActionDelegate: nil, urlOpener: urlOpener)
        
        let action = IterableAction.action(fromDictionary: ["type" : "openUrl", "data" : urlString])!
        _ = actionRunner.execute(action: action, from: .push)
        
        if #available(iOS 10.0, *) {
            XCTAssertNil(urlOpener.preIos10openedUrl)
            XCTAssertNil(urlOpener.ios10OpenedUrl)
        } else {
            XCTAssertNil(urlOpener.preIos10openedUrl)
            XCTAssertNil(urlOpener.ios10OpenedUrl)
        }
    }
    
    func testCustomAction() {
        let customActionDelegate = MockCustomActionDelegate(returnValue: false)
        let action = IterableAction.action(fromDictionary: ["type" : "customActionName"])!
        let actionRunner = IterableActionRunner(urlDelegate: nil, customActionDelegate: customActionDelegate, urlOpener: MockUrlOpener())
        _ = actionRunner.execute(action: action, from:.push)
        
        XCTAssertEqual(customActionDelegate.action, action)
    }
    
}
